internal/crypto/cms/signed.go

```diff
@@ -0,0 +1,247 @@
+	if cert == nil {
+		return nil, ErrCertificateNotFound
+	}
+
+	// verify signer certificate
+	if _, err := cert.Verify(*opts); err != nil {
+		return cert, VerificationError{Detail: err}
+	}
+
+	// verify signature
+	if err := d.verifySignature(signer, cert); err != nil {
+		return nil, err
+	}
+
+	// verify attribute
+	return cert, d.verifyAttributes(signer, cert)
+}
+
+// verifySignature verifies the signature with a trusted certificate.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifySignature(signer *SignerInfo, cert *x509.Certificate) error {
+	// verify signature
+	algorithm := oid.ToSignatureAlgorithm(
+		signer.DigestAlgorithm.Algorithm,
+		signer.SignatureAlgorithm.Algorithm,
+	)
+	if algorithm == x509.UnknownSignatureAlgorithm {
+		return VerificationError{Message: "unknown signature algorithm"}
+	}
+
+	signed := d.Content
+	if len(signer.SignedAttributes) > 0 {
+		encoded, err := asn1.MarshalWithParams(signer.SignedAttributes, "set")
+		if err != nil {
+			return VerificationError{Message: "invalid signed attributes", Detail: err}
+		}
+		signed = encoded
+	}
+
+	if err := cert.CheckSignature(algorithm, signed, signer.Signature); err != nil {
+		return VerificationError{Detail: err}
+	}
+	return nil
+}
+
+// verifyAttributes verifies the signed attributes.
+//
+// References:
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifyAttributes(signer *SignerInfo, cert *x509.Certificate) error {
+	// verify attributes if present
+	if len(signer.SignedAttributes) == 0 {
+		return nil
+	}
+
+	var contentType asn1.ObjectIdentifier
+	if err := signer.SignedAttributes.TryGet(oid.ContentType, &contentType); err != nil {
+		return VerificationError{Message: "invalid content type", Detail: err}
+	}
+	if !d.ContentType.Equal(contentType) {
+		return VerificationError{Message: "mismatch content type"}
+	}
+
+	var expectedDigest []byte
+	if err := signer.SignedAttributes.TryGet(oid.MessageDigest, &expectedDigest); err != nil {
+		return VerificationError{Message: "invalid message digest", Detail: err}
+	}
+	hash, ok := oid.ToHash(signer.DigestAlgorithm.Algorithm)
+	if !ok {
+		return VerificationError{Message: "unsupported digest algorithm"}
+	}
+	actualDigest, err := hashutil.ComputeHash(hash, d.Content)
+	if err != nil {
+		return VerificationError{Message: "hash failure", Detail: err}
+	}
+	if !bytes.Equal(expectedDigest, actualDigest) {
+		return VerificationError{Message: "mismatch message digest"}
+	}
+
+	// sanity check on signing time
+	var signingTime time.Time
+	if err := signer.SignedAttributes.TryGet(oid.SigningTime, &signingTime); err != nil {
+		if err == ErrAttributeNotFound {
+			return nil
+		}
+		return VerificationError{Message: "invalid signing time", Detail: err}
+	}
+	if signingTime.Before(cert.NotBefore) || signingTime.After(cert.NotAfter) {
```

Comment:
We are only checking `signingTime` against the leaf cert here. To be safer, we actually ought to check against the whole chain (in our case, that would be chain provided by the TSA).